####-release_cycle.md

```diff
@@ -0,0 +1,138 @@
+
+**out-of-scope**: 
+The starting date for the implementation of this RFC is, in principle, outside the scope of this document.
+
+## User Benefit
+
+The RFC aims to benefit users and Qiskit ecosystem developers, because they will have guaranties that their software will run for a defined period of time and they could plan the transition period. 
+
+## Design Proposal
+> This is the focus of the document. Explain the proposal from the perspective of
+> educating another user on the proposed features.
+> 
+> This generally means:
+> 
+> - Introducing new concepts and nomenclature
+> - Using examples to introduce new features
+> - Implementation and Migration path with associated concerns
+> - Communication of features and changes to users
+> 
+> Focus on giving an overview of impact of the proposed changes to the target
+> audience.
+> 
+> Factors to consider:
+> 
+> - Performance
+> - Dependencies
+> - Maintenance
+> - Compatibility
+
+Currently 
+
+```mermaid
+gantt
+    dateFormat  YYYY-MM-DD
+    tickInterval 3month
+    todayMarker off
+    axisFormat %b-%Y
+    0.17    :r017, 2021-04-01, 2021-07-12
+    0.18    :r018, after r017, 2021-12-06
+    0.19    :r019, after r018, 2022-03-31
+    0.20    :r020, after r019, 2022-06-30
+    0.21rc  :r021rc, 2022-06-23, 2022-06-30
+    0.21    :r021, after r020, 2022-10-13
+    0.22rc  :r022rc, 2022-10-07, 2022-10-13
+    0.22    :r022, after r021, 2023-01-26
+    0.23rc  :r023rc, 2023-01-16, 2023-01-26
+    0.23    :r024, after r022, 2023-05-04
+    0.24rc  :r024rc, 2023-04-20, 2023-05-04
+    0.24    :milestone, 2023-05-04
+```
+
+Suggest
+```mermaid
+gantt
+    dateFormat  YYYY-MM-DD
+    tickInterval 3month
+    todayMarker off
+    axisFormat %b
+    section X.*
+    X.0   :rX0Y0, 2024-01-31, 91d
+    X.1   :rX0Y1, after rX0Y0, 91d
+    X.2   :rX0Y2, after rX0Y1, 91d
+    X.3   :rX0Y3, after rX0Y2, 91d
+    X.3.*   :done, X0Y3s, after rX0Y3, 183d
+    section X+1.*
+    X+1.0    :rX1Y0, after rX0Y3, 91d
+    X+1.1    :rX1Y1, after rX1Y0, 91d
+    X+1.2    :rX1Y2, after rX1Y1, 91d
+    X+1.3    :rX1Y3, after rX1Y2, 91d
+    X+1.3.*  :done, rX1Y3s, after rX1Y3, 183d
+    section X+2.*
+    X+2.0    :rX2Y0, after rX1Y3, 91d
+    X+2.1    :rX2Y1, after rX2Y0, 91d
+    X+2.2    :rX2Y2, after rX2Y1, 91d
+    X+2.3    :rX2Y3, after rX2Y2, 91d
+    X+2.3.*  :done, rX2Y3s, after rX2Y3, 183d
+```
+
+## Detailed Design
+
+> Technical reference level design. Elaborate on details such as:
+> 
+> - Implementation procedure
+>   - If spans multiple projects cover these parts individually
+> - Interaction with other features
+> - Dissecting corner cases
+> - Reference definition, eg., formal definitions.
+
+
+### Deprecations
+
+* deprecation warnings can only occur 1+ minor releases after the replacement functionality is added
+* deprecation warnings must be emitted for a complete major version" (i.e. 1+ year)
```

Comment:
Added from https://github.com/Qiskit/RFCs/pull/34#issuecomment-1549428294 @jakelishman 

I would like to offer an alternative as this sounds like a lot of time to me. Introducing a deprecation warning in X.Y and not being able to remove it until X+2.0 is a huge time period of dangling code. If we keep X.3 support for extra 6 month, that should be enough to remove it in X+1.0.

Reply:
I see the trouble with this approach is that it assumes that every user will upgrade through every minor version, so will see deprecations introduced in minor releases.  To me, introducing major versions means that we expect that the required upgrade path is only at the granularity of "through each major version", and so my form is what guarantees that a user will see every deprecation warning.

Reply:
For example, currently our upgrade strategy is that you can't expect to go from Terra 0.22 to 0.24 without something breaking without warning - you must go via 0.23 to ensure you've seen all warnings and had the opportunity to smoothly upgrade.  I would foresee that if we add major versioning, the story ought to be "you must upgrade through each major, but you needn't install each minor".

Reply:
> it assumes that every user will upgrade through every minor version, so will see deprecations introduced in minor releases.

Actually, the necessary assumption is weaker: before upgrading to a new major version, the user would only need to use the _final_ minor version of the previous release, because there, all deprecation warnings necessary for upgrading will be visible and can be fixed.

Julia is an example that illustrates both the benefits and drawbacks of this approach.  Julia 0.7.0 and 1.0.0 were released within ~48 hours of each other.  The only difference between them is that 0.7.0 contained compatibility (and deprecation warnings) on code that worked under 0.6, while 1.0 removed all such support.  If there is ever a Julia 2.0, I expect the same thing to happen: users will be expected to use the "final" 1.X release and fix all deprecation warnings before trying the code under Julia 2.0.

The downside to this approach was that, in practice, people not intimately familiar with the release plan upgraded directly from 0.6.x to 1.0, and were met with a poor user experience: tons of code that worked on the previous release simply failed.  If the goal is to get users to use a certain release before upgrading, it really does need to be emphasized in _all_ the obvious locations, and then some.

Reply:
> people not intimately familiar with the release plan upgraded directly from 0.6.x to 1.0, and were met with a poor user experience: tons of code that worked on the previous release simply failed.

I see this as expected, as a major upgrade.

> I see the trouble with this approach is that it assumes that every user will upgrade through every minor version

Indeed. This is an assumption that needs to be explicit. However, it is not more obscure than the current one.
Downstream library should depend on the fairly standard `~=X.Y` or `==X.*`. With the extended period for `X.3.*` receiving critical bug fix support and not introducing new deprecations, pinning the previous major is always an option and gives a good period of time to migrate.

Reply:
My issue is that adding major versions into the mix does change the "obscurity" of the policy.  I expect that code I write for one version of a library with no warnings will work in the next version of that library (now maybe with warnings), at whatever granularity that library does it major releases.  Once we have majors, I see them as defining the required-upgrade granularity, as opposed to our current system where minors are that top level.  That's not a requirement of semantic versioning as a spec, but that's because that spec is only for associating the scope of changes with the version number, not for any of the development/upgrade strategies that go with it.

Requiring that the 2->3 upgrade goes via 2.last and heavily publicising that is one way of dealing with my concern, but I think we've seen that we won't ever reach all our users, and I'd argue that this is a much more "obscure" requirement.  People expect instability when something's versioned 0.x, but there's more expectation of stability when it's 1.x+.

Reply:
@garrison I actually really liked the Julia approach of having a final minor version to introduce all the deprecations prior to the major version bump. It provided a concrete path to follow as a developer to get to working code despite introducing a lot of change in 1.0.

In general, I think relying upon the _final_ minor version as a necessary step to expose all the deprecations is a very teachable system.

Reply:
The notion of "final minor release" was introduced in 3ec912ba0ae7e5664cdc03692f45ac9506caefea